<a href="https://colab.research.google.com/github/Jesus2342/Computer_vision/blob/proyecto_final/steer_angle_py.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense
from tensorflow.keras.preprocessing.image import ImageDataGenerator
import pandas as pd
import numpy as np
import cv2
import os
from sklearn.model_selection import train_test_split


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import os
DIR = "/content/drive/MyDrive/Colab Notebooks/Vision_computacional/deep_net/"
os.chdir(DIR)

In [ ]:

#Path del primer archivo csv
image_folder_path_1 ="/content/drive/MyDrive/Colab Notebooks/Vision_computacional/deep_net/DAY_1_Dataset"
os.chdir(image_folder_path_1)
csv_file_day_1 = "day_1_dataset_report.csv" #Nombre del primer archivo excel
dataset_day_1 = pd.read_csv(csv_file_day_1)

dataset_day_1.head(3)

image_name  steering_angle
0  2024-06-19 18-53-22.png            0.00
1  2024-06-19 18-53-27.png            0.08
2  2024-06-19 18-53-32.png            0.08

In [ ]:
print(image_folder_path_1)

/content/drive/MyDrive/Colab Notebooks/Vision_computacional/deep_net/DAY_1_Dataset


In [ ]:
#Estrategia: Las imagenes deben estandarizar una dimension de 36x18, ademas de contar con dos canales.

def estandarizar(imagen):
    img=cv2.imread(imagen)
    dim=img.shape

    #Obtener las dimensiones de la imagen en gris
    current_weight, current_height =img.shape[:2]

    #Acorde a nuestro patron de clasificacion, las imagenes deben medir 64x64

    std_weight=64
    std_height=64

    #Escenario si nuestra imagen tiene un alto y ancho diferente al especificado
    if current_height>std_height or current_weight>std_weight:
        #Actualizar la altura actual
        current_height=std_height
        #Actualizar el ancho actual
        current_weight=std_weight
        #Confirmar las dimensiones
        updated_dim=(current_height,current_weight)
        #Realizar el resize
        img=cv2.resize(img,updated_dim)
        #Normalizar
        img=img/255


    #Cambio a gris si hay 3 colores en la imagen
    #if (len(dim)==3):
     #   img_gray = cv2.cvtColor(img,cv2.COLOR_BGR2GRAY)

    #Si la imagen ya tiene dos calores se conserva sin hacer cambios
   # else:
    #    img_gray = img


    return img

In [75]:
imagenes=[]
angulos=[]

In [76]:
def iteracion_index_row(data_set_day, path_imagenes):
    for index, row in data_set_day.iterrows():
        img_path= os.path.join(path_imagenes,row['image_name'])
        img=estandarizar(img_path)
        imagenes.append(img)
        angulos.append(row['steering_angle'])

    return imagenes, angulos


In [77]:
lectura_imagenes_angulos_dataset_1=iteracion_index_row(dataset_day_1, image_folder_path_1)

In [78]:
print(f"Cantidad de imagenes: {len(imagenes)}")
print(f"Cantidad de angulos: {len(angulos)}")

Cantidad de imagenes: 562
Cantidad de angulos: 562


En este punto debe haber:

-Cantidad de imagenes: 562


-Cantidad de angulos: 562

In [79]:
#Path del segundo archivo archivo csv
image_folder_path_2 ="/content/drive/MyDrive/Colab Notebooks/Vision_computacional/deep_net/DAY_2_Dataset"
os.chdir(image_folder_path_2)
csv_file_day_2 = "day_2_dataset_report.csv" #Nombre del primer archivo excel
dataset_day_2 = pd.read_csv(csv_file_day_2)

dataset_day_2.head(3)

image_name  steering_angle
0  2024-06-20 17-19-05.png            0.00
1  2024-06-20 17-19-10.png            0.12
2  2024-06-20 17-19-15.png            0.12

In [80]:
lectura_imagenes_angulos_dataset_2=iteracion_index_row(dataset_day_2, image_folder_path_2)

In [81]:
print(f"Cantidad de imagenes: {len(imagenes)}")
print(f"Cantidad de angulos: {len(angulos)}")

Cantidad de imagenes: 1346
Cantidad de angulos: 1346


En este punto debe haber:

-Cantidad de imagenes: 1346


-Cantidad de angulos: 1346

In [83]:
#Path de dataset_3_1 archivo archivo csv
image_folder_path_3_1 ="/content/drive/MyDrive/Colab Notebooks/Vision_computacional/deep_net/DAY_3_Dataset_1"
os.chdir(image_folder_path_3_1)
csv_file_day_3_1 = "day_3_dataset_1_report.csv" #Nombre del tercer archivo excel
dataset_day_3_1 = pd.read_csv(csv_file_day_3_1)

dataset_day_3_1.head(3)

image_name  steering_angle
0  2024-06-21 16-33-50.png            0.00
1  2024-06-21 16-33-53.png            0.00
2  2024-06-21 16-33-56.png            0.08

In [85]:
lectura_imagenes_angulos_dataset_3_1=iteracion_index_row(dataset_day_3_1, image_folder_path_3_1)

In [86]:
print(f"Cantidad de imagenes: {len(imagenes)}")
print(f"Cantidad de angulos: {len(angulos)}")

Cantidad de imagenes: 2860
Cantidad de angulos: 2860


En este punto debe haber:

-Cantidad de imagenes: 2860


-Cantidad de angulos: 2860

In [87]:
#Path del dataset 3_2 archivo csv
image_folder_path_3_2 ="/content/drive/MyDrive/Colab Notebooks/Vision_computacional/deep_net/DAY_3_Dataset_2"
os.chdir(image_folder_path_3_2)
csv_file_day_3_2 = "day_3_dataset_2_report.csv" #Nombre del tercer archivo excel
dataset_day_3_2 = pd.read_csv(csv_file_day_3_2)

dataset_day_3_2.head(3)

image_name  steering_angle
0  2024-06-21 18-12-03.png             0.0
1  2024-06-21 18-12-06.png             0.0
2  2024-06-21 18-12-09.png             0.0

In [89]:
lectura_imagenes_angulos_dataset_3_2=iteracion_index_row(dataset_day_3_2, image_folder_path_3_2)

In [90]:
print(f"Cantidad de imagenes: {len(imagenes)}")
print(f"Cantidad de angulos: {len(angulos)}")

Cantidad de imagenes: 3886
Cantidad de angulos: 3886


En este punto debe haber:

-Cantidad de imagenes: 3886


-Cantidad de angulos: 3886

In [92]:
#Path del dataset 3_3 archivo csv
image_folder_path_3_3 ="/content/drive/MyDrive/Colab Notebooks/Vision_computacional/deep_net/DAY_3_Dataset_3"
os.chdir(image_folder_path_3_3)
csv_file_day_3_3 = "day_3_dataset_3_report.csv" #Nombre del tercer archivo excel
dataset_day_3_3 = pd.read_csv(csv_file_day_3_3)

dataset_day_3_3.head(3)

image_name  steering_angle
0  2024-06-22 00-14-04.png        0.000000
1  2024-06-22 00-14-05.png       -0.003781
2  2024-06-22 00-14-06.png       -0.109872

In [93]:
import os

def contar_archivos_png(folder_path):
    count = 0
    for filename in os.listdir(folder_path):
        if filename.lower().endswith('.png'):
            count += 1
    return count

# Ruta del folder que quieres verificar
folder_path = image_folder_path_3_3  # Reemplaza con la ruta a tu folder

cantidad_png = contar_archivos_png(folder_path)
print(f'Cantidad de archivos PNG en {folder_path}: {cantidad_png}')


Cantidad de archivos PNG en /content/drive/MyDrive/Colab Notebooks/Vision_computacional/deep_net/DAY_3_Dataset_3: 4478


In [94]:
lectura_imagenes_angulos_dataset_3_3=iteracion_index_row(dataset_day_3_3, image_folder_path_3_3)

In [95]:
print(f"Cantidad de imagenes: {len(imagenes)}")
print(f"Cantidad de angulos: {len(angulos)}")

Cantidad de imagenes: 8360
Cantidad de angulos: 8360


Comienza el entrenamiento

In [96]:
X = np.array(imagenes)
y = np.array(angulos)

In [97]:
#Division de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 42)


In [99]:
model = Sequential([
    Conv2D(32, (3, 3), activation='relu', input_shape=(64, 64, 3)),
    MaxPooling2D((2, 2)),
    Conv2D(64, (3, 3), activation='relu'),
    MaxPooling2D((2, 2)),
    Conv2D(128, (3, 3), activation='relu'),
    MaxPooling2D((2, 2)),
    Flatten(),
    Dense(128, activation='relu'),
    Dense(64, activation='relu'),
    Dense(1)  # regresión del angulo predicho
])

# Compilación del modelo
model.compile(optimizer='adam', loss='mean_squared_error', metrics=["accuracy"])

# Entrenamiento
model.fit(X_train, y_train, epochs=10,  batch_size=32)

# Evaluación del modelo
test_loss = model.evaluate(X_test, y_test)


Epoch 1/10
183/183 [==============================] - 39s 206ms/step - loss: 0.0113 - accuracy: 0.2418
Epoch 2/10
183/183 [==============================] - 37s 204ms/step - loss: 0.0102 - accuracy: 0.2418
Epoch 3/10
183/183 [==============================] - 36s 196ms/step - loss: 0.0098 - accuracy: 0.2418
Epoch 4/10
183/183 [==============================] - 38s 206ms/step - loss: 0.0094 - accuracy: 0.2418
Epoch 5/10
183/183 [==============================] - 36s 196ms/step - loss: 0.0092 - accuracy: 0.2418
Epoch 6/10
183/183 [==============================] - 36s 197ms/step - loss: 0.0089 - accuracy: 0.2418
Epoch 7/10
183/183 [==============================] - 38s 206ms/step - loss: 0.0085 - accuracy: 0.2418
Epoch 8/10
183/183 [==============================] - 36s 196ms/step - loss: 0.0083 - accuracy: 0.2418
Epoch 9/10
183/183 [==============================] - 36s 197ms/step - loss: 0.0081 - accuracy: 0.2418
Epoch 10/10
79/79 [==============================] - 4s 48ms/step - loss: